### PR TITLE
Initial import of the *Google Instrumentation Agent for Java*

### PR DESCRIPTION
**Pull request by ubschmidt2**
Additional background in b/33670885.

### CODE REVIEW

File: agent/README.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Google Instrumentation Agent for Java
+
+[![Build Status][travis-image]][travis-url] [![Maven Central][maven-image]][maven-url]
+
+The *Google Instrumentation Agent for Java* collects and sends latency data
+about your Java process to Stackdriver Trace for analysis and visualization
```
Comment (adriancole) [inline agent/README.md]: nit.. need to wash some of this stuff for the opencensus branding. Ex preferring wording like "opencensus backends such as StackDriver" vs being the only target.

File: agent/src/main/java/com/google/instrumentation/agent/AgentMain.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    logger.fine("Initialized.");
+  }
+
+  private static class LazyLoaded {
+
+    // Avoid triggering Errorprone's warning about using {@link Class#forName} on a
+    // class that is on the compile-time classpath. Though the class is on the compile-time
+    // classpath, it may be missing at run time, in which case the bytecode will not be instrumented
+    // for automatic context progagation.
+    private static final String CONTEXT_CLASS_NAME = "io.grpc.".toLowerCase(Locale.US) + "Context";
```
Comment (adriancole) [inline agent/src/main/java/com/google/instrumentation/agent/AgentMain.java]: is this toLC stuff a bit overkill? or are you expecting something interesting here?

File: agent/src/main/java/com/google/instrumentation/agent/AgentMain.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        agentBuilder = agentBuilder
+                .type(ExecutorInstrumentation.matcher())
+                .transform(ExecutorInstrumentation.transformer());
+
+        agentBuilder = agentBuilder
+                .type(ThreadInstrumentation.matcher())
+                .transform(ThreadInstrumentation.transformer());
+      } catch (ClassNotFoundException e) {
+        logger.log(Level.FINE, "Could not load io.grpc.Context, therefore not instrumenting "
```
Comment (adriancole) [inline agent/src/main/java/com/google/instrumentation/agent/AgentMain.java]: should this reuse the constant for the class name?

File: agent/src/main/java/com/google/instrumentation/agent/ExecutorInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.concurrent.Executor;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.utility.JavaModule;
+
+/**
+ * Propagates the context of the caller of {@link Executor#execute} to the submitted
```
Comment (adriancole) [inline agent/src/main/java/com/google/instrumentation/agent/ExecutorInstrumentation.java]: should this defend against double-wrapping? ex if the app already did this

File: agent/src/main/java/com/google/instrumentation/agent/ThreadInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.utility.JavaModule;
+
+/**
+ * Propagates the context of the caller of {@link Thread#start} to the new thread, just like the
```
Comment (adriancole) [inline agent/src/main/java/com/google/instrumentation/agent/ThreadInstrumentation.java]: double-duty when the executor starts a new thread implicitly (ex from thread pool growing)? should there be a caution here? if not why?

File: agent/src/main/java/com/google/instrumentation/agent/bootstrap/ContextProxy.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ * can be used from code that the agents injects into classes loaded by the bootstrap classloader.
+ * To avoid having to also load {@link io.grpc.Context} from the bootstrap classloader,
+ * {@link ContextProxy} delegates to its implementation
+ * {@link com.google.instrumentation.agent.ContextProxyImpl}, which is loaded by the system
+ * classloader. {@link ContextProxy} cannot directly reference {@link io.grpc.Context}, thus Object
+ * is used in the interface, whereas its implementation
+ * ({@link com.google.instrumentation.agent.ContextProxyImpl}) can downcast to
+ * {@link io.grpc.Context} whenever required.
+ */
+public abstract class ContextProxy {
```
Comment (adriancole) [inline agent/src/main/java/com/google/instrumentation/agent/bootstrap/ContextProxy.java]: in general, curious if we shouldn't document more the rationale somewhere about directly working with io.grpc.Context. While true that census holds this assumption now, it isn't mentioned directly (unless I missed something). So, maybe at least in comments discuss this is an alternative to proxying something like ContextUtils

File: agent/src/main/java/com/google/instrumentation/agent/bootstrap/package-info.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2017, Google Inc.
```
Comment (adriancole) [inline agent/src/main/java/com/google/instrumentation/agent/bootstrap/package-info.java]: should we start doing things like guava? ex Copyright the OpenCensus authors?

File: agent/src/test/java/com/google/instrumentation/agent/ExecutorInstrumentationTest.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+/** Integration tests for {@link ExecutorInstrumentation}. */
+public class ExecutorInstrumentationTest {
+
+  private static final Context.Key<String> KEY = Context.key("mykey");
+
+  private ExecutorService executor;
```
Comment (adriancole) [inline agent/src/test/java/com/google/instrumentation/agent/ExecutorInstrumentationTest.java]: hmm are these just testing Context? is this sanity checking usage patterns, or is something else going on?

Comment (adriancole) [inline agent/src/test/java/com/google/instrumentation/agent/ExecutorInstrumentationTest.java]: if something is setting up bytebuddy that isn't obvious.. maybe make a comment or javadoc explicitly where it is?

**Comment by ubschmidt2** [comment]
Thanks for the review comments!

FYI, I'm splitting this PR into smaller, more trivial bites.

File: agent/README.md (hunk 1 above)
Comment (ubschmidt2) [inline agent/README.md]: This is now part of https://github.com/census-instrumentation/instrumentation-java/pull/364, using the OpenCensus branding.

File: agent/src/main/java/com/google/instrumentation/agent/bootstrap/package-info.java (hunk 7 above)
Comment (ubschmidt2) [inline agent/src/main/java/com/google/instrumentation/agent/bootstrap/package-info.java]: Will clarify. Leaving this as is for now.

File: agent/src/main/java/com/google/instrumentation/agent/ExecutorInstrumentation.java (hunk 4 above)
Comment (ubschmidt2) [inline agent/src/main/java/com/google/instrumentation/agent/ExecutorInstrumentation.java]: Moved to https://github.com/census-instrumentation/opencensus-java/pull/411. Added a TODO there.

File: agent/src/main/java/com/google/instrumentation/agent/bootstrap/ContextProxy.java (hunk 6 above)
Comment (ubschmidt2) [inline agent/src/main/java/com/google/instrumentation/agent/bootstrap/ContextProxy.java]: Moved to https://github.com/census-instrumentation/opencensus-java/pull/411.
Introduced a ContextStrategy interface. Only implementation for now is GrpcContextStrategy. Could add alternatives later.

**Comment by ubschmidt2** [comment]
I'm closing this PR as it is mostly obsolete.
https://github.com/census-instrumentation/opencensus-java/pull/411 contains the instrumentation for Executor. I'll send a PR for the remaining code once #411 is merged.

**Pull request closed without merging**